torch/_inductor/ir.py

```diff
@@ -4015,3 +4015,106 @@ def debug_str(self, name="block"):
             "",
             code.strip().replace("def forward(", f"def {name}("),
         )
+
+
+class Wait(ExternKernel):
+    """
+    Wait should not be used by itself.  It should always be constructed in tandem
+    with a collective op that produces a work to wait on.
+    """
+
+    def __init__(
+        self,
+        layout,
+        inputs,
+        constant_args=(),
+    ):
+        super().__init__(None, layout, inputs, constant_args)
+        self.name = V.graph.register_buffer(self)
+
+    def should_allocate(self):
+        return False
+
+    def codegen(self, wrapper):
+        (all_reduce,) = [t.codegen_reference() for t in self.inputs]
+        work = f"{all_reduce}_work"  # hacky way to name work objs..
+        wrapper.writeline(f"{work}.wait()")
+
+        # largely symbolic from the runtime perspective, but we must ensure from scheduler
+        # perspective that Wait's buffer is the one being consumed by consumers, to ensure
+        # wait op gets scheduled at all.  And as a side effect, wait's buffer name needs to actually
+        # point to something in the codegen output
+
+        # this 'del' is screwing up scheduler 'reuse' of this buf. which i'm not sure should even be allowed
+        # wrapper.writeline(f"{self.get_name()} = {all_reduce}; del {all_reduce}")
+        wrapper.writeline(f"{self.get_name()} = {all_reduce}")
```

Comment:
You might need to add this to `get_alias_names()` to prevent the other value from being reused.

Reply:
Could we just capture wait as part of tracing and make it functional, IE "wait(tensor) -> tensor".

Even better if we can capture it through AsyncCommTensor so it already lands on the right spot.

This way any code motion/scheduling pass would just work OOTB as we'd have captured the dependencies.

Reply:
well, we don't gain or lose information by having wait in our initial traced IR.

the question is, where do we want to do the scheduling?  so far, the plan from the discussions with folks is to try to do the scheduling in inductor.  sync with @lessw2020 also on this.

Reply:
ok- we've changed to do things the way proposed by @kumpera, `Wait` is no longer added automatically by Inductor and is lowered separately from the input FX graph.